extensions/ql-vscode/src/pure/messages.ts

```diff
@@ -1018,6 +1063,16 @@ export const compileUpgrade = new rpc.RequestType<WithProgressId<CompileUpgradeP
  */
 export const compileUpgradeSequence = new rpc.RequestType<WithProgressId<CompileUpgradeSequenceParams>, CompileUpgradeSequenceResult, void, void>('compilation/compileUpgradeSequence');
 
+/**
+ * Start a new structured log in the evaluator, terminating the previous one if it exists
```

Comment:
Does this mean we can only have one structured log at the same time? What if we have multiple queries running? I guess there's no way to separate logs for each of them.

Reply:
Edoardo and I discussed this this morning — it sounds like if we have multiple queries running at the same time, the interleaved messages will write to the log file for the most recent query. This is also what happens to existing logs so we thought this would be acceptable for the time being.

Reply:
OK...At the time we wrote the interleaved  log messages, there wasn't an easy way to distinguish messages for different queries.

 With structured logging, in theory, it should be easier to add a unique identifier that distinguishes where the query came from. 

When we invoke a query run, the extension passes in an id. See: https://github.com/github/vscode-codeql/blob/39f9c082b988610cf291df1d5e1eecbab845e7d4/extensions/ql-vscode/src/run-queries.ts#L123-L132

This id is used for callbacks, cancellation, and progress. Is there any possibility of adding the id to the structure log message? This is not something for now, but some future work.

Reply:
Indeed, as I discussed with Angela I don't think we can do any better than this (and already don't for the unstructured logs). The problem is that from the evaluator's point of view there isn't really a concept of what evaluations are being done for which query. When we start a query we just enqueue the whole dependency graph of predicates that it needs to evaluate that query, and that can partially overlap with other already running queries.

Reply:
Interesting, so in theory, if predicates are shared across queries, the same work can be done for both of them?

Reply:
Yes, I believe within a single run of the query server, the more recent queries can pull from some cache, but please correct me if incorrect @edoardopirovano !

Reply:
Concretely in the evaluator what happens is that the shared predicate will become a single node in the evaluation graph that is needed by both queries. Then, it's evaluated once and the result is used by both queries wherever it is needed (directly in the query, or in one of it's transitive dependencies).

We do track demand in the evaluation graph, so for many predicates we can associate a query that most recently created demand for it in the `queryCaysingWork` field of the predicate start event. However, this is a little misleading because often this predicate would've been needed by other queries, so it isn't particularly meaningful to associate the time spend evaluating it to one query.

Rather than trying to achieve something similar to what we do with the cancellation tokens for the logs, I think I would much prefer to have a warning that running queries in parallel on the same DB will produce unreliable logs and timing information. Perhaps the extension could even have an option to queue queries for sequential running rather than running them in parallel, for users who care about getting accurate benchmarking information.